src/handler/core.rs

```diff
@@ -108,12 +162,47 @@ where
     pub fn chain(self, next: Self) -> Self {
         let required_update_kinds_set = self.description().merge_chain(next.description());
 
-        from_fn_with_description(required_update_kinds_set, move |event, cont| {
-            let this = self.clone();
-            let next = next.clone();
-
-            this.execute(event, |event| next.execute(event, cont))
-        })
+        let new_sig = match (&self.data.sig, &next.data.sig) {
```

Comment:
I'm a bit iffy about doing this here, could we maybe do all the type-checking via the descriptions (which should really have been named metadata, imo...)?

Reply:
The reason we type-check not via descriptions is because we want to _always_ type-check handlers and still use descriptions. For example, in `teloxide` we're using `DpHandlerDescription`; to be able to type-check, we would need to somehow make the handler polymorphic over any description so as to instantiate the same handler with both `DpHandlerDescription` and "`TypeCheckDescription`".

Reply:
The descriptions are trivially composable though, you can include `TypeCheckDescription` inside `DpHandlerDescription` (just like we already use some description inside `DpHandlerDescription`, I don't remember the name tho...)

Reply:
But that would require a user to compose their description types with our `TypeCheckDescription` instead of type-checking automatically?

Reply:
Yes, although I'm not sure if it's a big deal?

We could also store the type checking description/meta always, e.g. `description: WithTypeCheck<Descr>`.

Reply:
I just think that type checking is more "inherent" to `Handler` than a typical description. For example, we explicitly construct `HandlerSignature` independently of descriptions in `endpoint`, `filter`, `inspect`, etc., because that requires access to `Injectable`'s generics that `HandlerDescription` simply doesn't have access to. Maybe we could somehow circumvent such limitations but I don't think it'd result in a simpler structure...

Reply:
Yeah okay, it's fine.